README.md
# cookie-boss

Takes [cookie-manager](https://github.com/juji/cookie-manager) (extends prototype) and [cookie](https://github.com/jshttp/cookie) (adds parse and serialize methods as class and prototype methods), and adds a toJSON method. It is the boss!

## Installation
```
npm install cookie-boss
```

## Usage
```javascript
const CookieBoss = require('cookie-boss');
const cookieboss = new CookieBoss();
const url = 'https://github.com'
const cookiesToStore = [
    `_octo=GH1.1.1249112545.1582225594; Path=/; Domain=github.com; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Secure`,
    `logged_in=no; Path=/; Domain=github.com; Expires=Sat, 20 Feb 2021 19:06:34 GMT; HttpOnly; Secure`
];

cookieboss.store(url, cookiesToStore);

const prepared = cookieboss.prepare(url);
// '_octo=GH1.1.1249112545.1582225594; logged_in=no'

const parsed = CookieBoss.parse(prepared);
// { _octo: 'GH1.1.1249112545.1582225594', logged_in: 'no' }

const json = cookieboss.toJSON();
// {
//     "github.com": [
//         "_octo=GH1.1.1249112545.1582225594; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Max-Age=31611923; Path=/; Domain=github.com; secure",
//         "logged_in=no; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Max-Age=31611923; Path=/; Domain=github.com; secure; HttpOnly"
//     ]
// }
```

Typescript supported too!

CHANGELOG.md
# Changelog

## [1.0.1] - 2019-02-22
### Fixed
- A bug when storing invalid cookies would cause a "Maximum call stack size exceeded" error when calling toJSON method.

### Changed
- Made "store" prototype method safer from storing bad data.

### Added
- "fromJSON" prototype method.
- "getCookies" prototype method.

## [1.1.0] - 2019-02-22
### Added
- Typescript method definitions.

tests/main.test.js
const year = new Date().getFullYear();
const CookieBoss = require('../');
const cookieboss = new CookieBoss();

it('Stores properly', () => {
    const cookieboss = new CookieBoss();
    const url = 'https://github.com'
    cookieboss.store(url, [
        `_octo=GH1.1.1249112545.1582225594; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; Secure`,
        `logged_in=no; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; HttpOnly; Secure`
    ]);
    
    const prepared = cookieboss.prepare(url);
    
    expect(prepared).toBe('_octo=GH1.1.1249112545.1582225594; logged_in=no');
});

it('Prepares proper JSON', () => {
    const cookieboss = new CookieBoss();
    const url = 'https://github.com'
    cookieboss.store(url, [
        `_octo=GH1.1.1249112545.1582225594; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; Secure`,
        `logged_in=no; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; HttpOnly; Secure`
    ]);
    const json = cookieboss.toJSON();
    
    expect(Object.keys(json)).toEqual(['github.com']);
    expect(json['github.com'].length).toBe(2);
});

it('Parses a cookie', () => {
    const cookieboss = new CookieBoss();
    const url = 'https://github.com'
    cookieboss.store(url, [
        `_octo=GH1.1.1249112545.1582225594; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; Secure`,
        `logged_in=no; Path=/; Domain=github.com; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; HttpOnly; Secure`
    ]);
    const cookie = '_octo=GH1.1.1249112545.1582225594; logged_in=no';
    const parsed = cookieboss.parse(cookie);
    
    expect(parsed).toEqual({
        _octo: 'GH1.1.1249112545.1582225594',
        logged_in: 'no'
    });
});

it('Throws an error when attempting to store a cookie with a given full url', () => {
    const cookieboss = new CookieBoss();
    cookieboss.store('badcookie', ['badcookie=yes; Path=/;']);
    
    expect(cookieboss.prepare('badcookie')).toBe('');
});

it('Stores cookie when domain is given in cookie', () => {
    const cookieboss = new CookieBoss();
    cookieboss.store('badcookie', ['badcookie=no; Path=/; Domain=goodcookie.com; Secure']);
    
    expect(cookieboss.prepare('https://goodcookie.com')).toBe('badcookie=no');
});

it('Stores cookie using cookie domain', () => {
    const cookieboss = new CookieBoss();
    cookieboss.store('.domain.com', [
        `user-id=123; Expires=Wed, 20 May ${year + 1} 18:42:48 GMT; Max-Age=7624700; Path=/; Domain=.domain.com; HttpOnly`
    ]);
    
    expect(cookieboss.prepare('https://domain.com')).toBe('user-id=123');
});

it('Stores cookie using cookies as a string', () => {
    const cookieboss = new CookieBoss();
    cookieboss.store('https://domain.com', `user-id=123; Expires=Wed, 20 May ${year + 1} 18:42:48 GMT; Max-Age=7624700; Path=/; Domain=.domain.com; HttpOnly`);
    
    expect(cookieboss.prepare('https://domain.com')).toBe('user-id=123');
});

it('Ignores empty domain', () => {
    const cookieboss = new CookieBoss();
    cookieboss.store('', [
        `nothing=nothing; Expires=Wed, 20 May ${year + 1} 18:42:48 GMT; Max-Age=7624700; Path=/; Domain=.domain.com; HttpOnly`
    ]);
    
    expect(cookieboss.prepare('')).toBe('');
});

it('Stores from JSON', () => {
    const cookieboss = new CookieBoss();
    
    cookieboss.fromJSON({
        "github.com": [
            `_octo=GH1.1.1249112545.1582225594; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; Max-Age=31467234; Path=/; Domain=github.com; secure`,
            `logged_in=no; Expires=Sat, 20 Feb ${year + 1} 19:06:34 GMT; Max-Age=31467234; Path=/; Domain=github.com; secure; HttpOnly`
        ]
    });
    
    expect(cookieboss.getCookies().length).toBe(2);
});


index.js
const SimpleCookie = require('simple-cookie');
const CookieManager = require('cookie-manager');
const Cookie = require('cookie');
const util = require('util');
const urlParse = require('url');

function CookieBoss() {
    // it doesn't take any arguments
    CookieManager.call(this);
}

util.inherits(CookieBoss, CookieManager);

CookieBoss.parse = Cookie.parse;
CookieBoss.serialize = Cookie.serialize;
CookieBoss.prototype.parse = Cookie.parse;
CookieBoss.prototype.serialize = Cookie.serialize;

CookieBoss.prototype.getCookies = function() {
    const getDomainCookies = (obj) => {
        /*
        {
            expires: 2027-10-28T05:00:39.000Z,
            httponly: false,
            secure: false,
            path: '/',
            domain: 'example.com',
            name: 'example',
            value: 'xxxxxxx',
            pathReg: /^\//
        }
        */
        return Object.values(obj);
    };
    const reduceFlatten = (total, current) => {
        if (Array.isArray(current)) {
            return total.concat(current.reduce(reduceFlatten, []));
        }
        
        return total.concat(current);
    };
    
    // build an array of all cookies from this.list
    return Object.values(this.list)
        // this will create an array of cookies
        .map(getDomainCookies)
        // flatten them down
        .reduce(reduceFlatten, []);
};

CookieBoss.prototype.toJSON = function() {
    return this.getCookies()
        // filter out expired cookies
        .filter((cookie) => {
            return Boolean(
                // cookie never expires
                !cookie.expires ||
                // cookie expires date is before now
                new Date().getTime() < cookie.expires
            );
        })
        // reduce them by domain
        .reduce((json, cookie) => {
            if (!json[cookie.domain]) {
                // create a new array for domain
                // so that we can add to it
                json[cookie.domain] = [];
            }
            
            // push stringified cookie to array
            json[cookie.domain].push(
                SimpleCookie.stringify(cookie)
            );
            
            return json;
        }, {});
};

CookieBoss.prototype.fromJSON = function(json) {
    Object.entries(json).forEach(([domain, cookies]) => {
        this.store(domain, cookies);
    });
};

CookieBoss.prototype.store = function(url, cookieStrs) {
    if (typeof cookieStrs === 'string') {
        cookieStrs = [cookieStrs];
    }
    
    if (!url || !Array.isArray(cookieStrs)) {
        return;
    }
    
    const { pathname, hostname } = urlParse.parse(url);
    
    Object.values(cookieStrs)
        // map the strings into cookie objects
        .map((cookieStr) => {
            // this will work with either full ur;'s or just the domain name
            const cookie = SimpleCookie.parse(cookieStr, pathname, hostname || url);
            
            cookie.pathReg = new RegExp('^' + cookie.path);
            
            return cookie;
        })
        // then assign each cookie to the manager
        .forEach((cookie) => {
            // domain for this cookie already exists
            if (this.domains.includes(cookie.domain)) {
                this.list[cookie.domain][cookie.name] = cookie;
            } else{
                this.list[cookie.domain] = {};
                this.list[cookie.domain][cookie.name] = cookie;
                
                const pattern = new RegExp(
                    cookie.domain.match(/^\./) ?
                        cookie.domain + '$' :
                        '^' + cookie.domain + '$'
                );
                
                this.domainReg.push(pattern);
                this.domains.push(cookie.domain);
            }
        });
    
    // calculate length
    this.length = this.getCookies().length;
};

module.exports = CookieBoss;

index.d.ts
export as namespace cookieboss;

export = CookieBoss;

type Cookie = {
    expires: Date,
    httponly: boolean,
    secure: boolean,
    path: string,
    domain: string,
    name: string,
    value: string,
    pathReg: RegExp
}

type CookieJSON = {
    [domain: string]: string | string[]
}

declare class CookieBoss {
    constructor();
    
    toJSON(): CookieJSON;
    fromJSON(json: CookieJSON): void;
    getCookies(): Cookie[];
    parse(cookieStr: string): CookieJSON;
    serialize(name: string, value: string, options: {
        domain?: string,
        encode?: (str: string) => string,
        expires?: Date,
        httpOnly?: boolean,
        maxAge?: number,
        sameSite?: true | false | 'lax' | 'none' | 'strict',
        secure?: number
    }): string;
    store(url: string, cookieStr: string | string[]): void;
    prepare(url: string): string;
}


examples/github.js
const CookieBoss = require('../');
const cookieboss = new CookieBoss();
const url = 'https://github.com'
const cookiesToStore = [
    `_octo=GH1.1.1249112545.1582225594; Path=/; Domain=github.com; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Secure`,
    `logged_in=no; Path=/; Domain=github.com; Expires=Sat, 20 Feb 2021 19:06:34 GMT; HttpOnly; Secure`
];

cookieboss.store(url, cookiesToStore);

const prepared = cookieboss.prepare(url);
// '_octo=GH1.1.1249112545.1582225594; logged_in=no'

const parsed = cookieboss.parse(prepared);
// { _octo: 'GH1.1.1249112545.1582225594', logged_in: 'no' }

const json = cookieboss.toJSON();
// {
//     "github.com": [
//         "_octo=GH1.1.1249112545.1582225594; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Max-Age=31611923; Path=/; Domain=github.com; secure",
//         "logged_in=no; Expires=Sat, 20 Feb 2021 19:06:34 GMT; Max-Age=31611923; Path=/; Domain=github.com; secure; HttpOnly"
//     ]
// }